gr-analog/lib/agc3_cc_impl.cc

```diff
@@ -40,97 +40,135 @@ agc3_cc_impl::agc3_cc_impl(float attack_rate,
     : sync_block("agc3_cc",
                  io_signature::make(1, 1, sizeof(gr_complex)),
                  io_signature::make(1, 1, sizeof(gr_complex))),
-      d_attack(attack_rate),
-      d_decay(decay_rate),
-      d_reference(reference),
-      d_gain(gain),
-      d_max_gain(65536),
-      d_reset(true),
-      d_iir_update_decim(iir_update_decim)
+      d_reset(true)
 {
+    set_reference(reference);
+    set_attack_rate(attack_rate);
+    set_decay_rate(decay_rate);
+    set_gain(gain);
+    set_max_gain(65536);
+    test_and_log_value_domain(iir_update_decim, "input power sampling stride");
+    d_iir_update_decim = iir_update_decim;
     set_output_multiple(iir_update_decim * 4);
     const int alignment_multiple = volk_get_alignment() / sizeof(gr_complex);
     set_alignment(std::max(1, alignment_multiple));
 }
 
 agc3_cc_impl::~agc3_cc_impl() {}
 
+void agc3_cc_impl::test_and_log_value_domain(float value, std::string_view description)
+{
+    if (value <= 0.0f) {
+        d_logger->error(
+            "Can't set {} power that is not strictly positive: {:g}", description, value);
+        throw std::domain_error(fmt::format("non-positive {}", description));
+    }
+}
+
+float agc3_cc_impl::attack_rate() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_attack;
+}
+float agc3_cc_impl::decay_rate() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_decay;
+}
+float agc3_cc_impl::reference() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_reference;
+}
+float agc3_cc_impl::gain() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_gain;
+}
+float agc3_cc_impl::max_gain() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_max_gain;
+}
+
+void agc3_cc_impl::set_attack_rate(float rate)
+{
+    test_and_log_value_domain(rate, "attack rate");
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_attack = rate;
+}
+void agc3_cc_impl::set_decay_rate(float rate)
+{
+    test_and_log_value_domain(rate, "decay rate");
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_decay = rate;
+}
+void agc3_cc_impl::set_reference(float reference)
+{
+    test_and_log_value_domain(reference, "reference");
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_reference = reference;
+}
+void agc3_cc_impl::set_gain(float gain)
+{
+    test_and_log_value_domain(gain, "gain");
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_gain = gain;
+}
+void agc3_cc_impl::set_max_gain(float max_gain)
+{
+    test_and_log_value_domain(max_gain, "maximum gain");
```

Comment:
This would block `max_gain = 0` (for unlimited), wouldn't it?

Reply:
I guess there should at least be a test that `max_gain = 0` is accepted, and ideally that it behaves as expected.